packages/functions/src/donor/DonorSwapAppointmentHandler.ts

```diff
@@ -0,0 +1,58 @@
+import { FunctionsApi } from "@zm-blood-components/common";
+
+import {
+  bookAppointment,
+  validateBookAppointment,
+} from "../common/BookAppointmentHelper";
+import {
+  cancelAppointment,
+  validateCancelAppointment,
+} from "../common/CancelAppointmentHelper";
+
+export default async function (
+  request: FunctionsApi.SwapAppointmentRequest,
+  callerId: string
+): Promise<FunctionsApi.SwapAppointmentResponse> {
+  const bookAppointmentValidation = validateBookAppointment(
+    request.bookAppointmentIds,
+    callerId
+  );
+
+  const cancelAppointmentValidation = validateCancelAppointment(
+    request.cancelAppointmentId,
+    callerId
+  );
+
+  // Unified promises due to low performance
+  const [
+    { status: bookAppointmentStatus },
+    { isValid: isCancelValid, invalidReason: cancellationInvalidReason },
+  ] = await Promise.all([
```

Comment:
If we ever decide to validate the donor doesn't already have a pending appointment, this will fail.

I think we should first cancel the appointment, and only after that book the other one

Reply:
The low performance is probably not due to this, but due to the cold start of the functions.

If you run this 3 times in a row, how long does it take to run?

Reply:
1. Should we not change the structure of the function then?
2. Currently I run both validations simultaneously and then book and then cancel, 
with an internal validation for each.
This means running 4 validations. 
If we should validate + cancel and then validate + book, why not skip the external validation altogether?
3. Running the validations simultaneously saved 600ms over synchronously,
running the tests repeatedly did not yield significant improvements to performance.

Reply:
Function now Cancels, then books the appointments.

Reply:
> If we ever decide to validate the donor doesn't already have a pending appointment, this will fail.

Then we can extract that booking status from the book validation function and uniquely pass it as acceptable,
only in the case of SWAP.